File: PAT_Basic_Level_by_c/1037 在霍格沃茨找零钱/1037 在霍格沃茨找零钱.c
/*
 * 题目：1037 在霍格沃茨找零钱
 * 实现原理：
 *   霍格沃茨货币单位：1 Galleon = 17 Sickles，1 Sickle = 29 Knuts
 *   计算应找回的零钱（应付金额 - 实付金额）。
 *   算法步骤：
 *   1. 将两个金额转换为Knuts总数
 *   2. 计算差值diff = total2 - total1
 *   3. 处理负数情况（输出负号）
 *   4. 将diff转换回Galleon.Sickle.Knut格式
 *   使用long long类型避免溢出。
 */
#include <stdio.h>

int main() {
    int g1, s1, k1, g2, s2, k2;  // 实付、应付金额的 Galleon.Sickle.Knut
    scanf("%d.%d.%d %d.%d.%d", &g1, &s1, &k1, &g2, &s2, &k2);
    long long total1 = g1 * 17 * 29 + s1 * 29 + k1;  // 实付金额统一换算为 Knut
    long long total2 = g2 * 17 * 29 + s2 * 29 + k2;  // 应付金额统一换算为 Knut
    long long diff = total2 - total1;  // 应找零（Knut 单位）
    if (diff < 0) {  // 应付大于实付 → 找零为负，先输出负号再取绝对值
        printf("-");
        diff = -diff;
    }
    int g = diff / (17 * 29);  // 换算回 Galleon 数
    diff %= 17 * 29;
    int s = diff / 29;  // 剩余的 Sickle 数
    int k = diff % 29;  // 最后的 Knut 数
    printf("%d.%d.%d\n", g, s, k);
    return 0;
}

File: PAT_Basic_Level_by_java/1037 在霍格沃茨找零钱/1037 在霍格沃茨找零钱.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * 1037 在霍格沃茨找零钱
 *
 * 题目描述：
 * 计算两个霍格沃茨货币值之差，货币单位为 Galleon.Sickle.Knut（加隆·西可·纳特），
 * 进制关系为：17 Sickle = 1 Galleon，29 Knut = 1 Sickle。
 *
 * 实现原理：
 * 1. 先将两个货币值统一转换为最小单位 Knut（纳特）：
 *    totalKnuts = G * 17 * 29 + S * 29 + K
 * 2. 计算差值（被减数 - 减数），若结果为负则记录符号位。
 * 3. 将差值的绝对值从 Knut 反向转换回 Galleon.Sickle.Knut 格式：
 *    - G = diff / (17 * 29)
 *    - S = (diff % (17 * 29)) / 29
 *    - K = diff % 29
 * 4. 若差值为负，在输出前加负号。
 *
 * 时间复杂度：O(1)，仅进行常数次算术运算
 */
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String p = br.readLine().trim();
        String p2 = br.readLine().trim();
        long v1 = toKnuts(p);
        long v2 = toKnuts(p2);
        long diff = v1 - v2;
        boolean neg = diff < 0;
        if (neg) diff = -diff;
        long g = diff / (17 * 29);
        diff %= (17 * 29);
        long s = diff / 29;
        long k = diff % 29;
        if (neg) System.out.print("-");
        System.out.println(g + "." + s + "." + k);
    }

    // 将 Galleon.Sickle.Knut 格式的字符串统一转换为最小单位 Knut
    private static long toKnuts(String str) {
        String[] parts = str.split("\\.");
        long g = Long.parseLong(parts[0]);
        long s = Long.parseLong(parts[1]);
        long k = Long.parseLong(parts[2]);
        return g * 17 * 29 + s * 29 + k;
    }
}
